docs_src/design/adr/device-service/0012-DeviceService-Filters.md

```diff
@@ -0,0 +1,152 @@
+
+### Design / Architecture
+
+A new function interface shall be defined that, when implemented, performs a Reading Filter operation.  A ReadingFilter function would take a parameter (an `Event` containing readings), check whether the `Readings` of the `Event` match on the filtering configuration (see below) and if they do then remove them from the `Event`.  The ReadingFilter function would return the `Event` object (minus filtered `Readings`) or `nil` if the `Event` held no more `Readings`.  Pseudo code for the generic function is provided below.
+
+``` go
+func (f Filter) ReadingFilter(lc logger.LoggingClient, event *models.Event) (*models.Event, error) {
+    // depending on impl; filtering for values in/out of a range, >, <, =, same, not same, from a particular name (device resource), etc.
+    if (len(event.Reading )) > 0)
+        return event
+    else
+        return nil
+}
+```
+
+Based on current needs/use cases, implementations of the function interface could include the following filter functions:
+
+``` go
+func (f Filter) FilterByValueInRange (lc logger.LoggingClient, event *models.Event) (*models.Event, error) {}
+
+func (f Filter) FilterByValueOutRange (lc logger.LoggingClient, event *models.Event) (*models.Event, error) {}
+
+func (f Filter) FilterByValueEqual (lc logger.LoggingClient, event *models.Event) (*models.Event, error) {}
+
+func (f Filter) FilterByValueNotEqual (lc logger.LoggingClient, event *models.Event) (*models.Event, error) {}
+
+func (f Filter) FilterByResourceNamesMatch (lc logger.LoggingClient, event *models.Event) (*models.Event, error) {}
+```
+
+!!! Note
+    The app functions SDK comes with `FilterByDeviceName` and `FilterByResourceName` functions today. The FilterByResourceName would behave similarly to FilterByResourceNameMatch.
+
+    The Filter structure houses the configuration parameters for which the filter functions work and filter on.
+
+A NewFilter function creates, initializes and returns a new instance of the filter based on the configuration provided.
+
+``` go
+func NewReadingNameFilter(filterValues []string) Filter {
+    return Filter{FilterValues: filterValues}
+}
+```
+
+### Sharing filter functions
+
+If one were to explore the filtering functions in the app functions SDK [filter.go](https://github.com/edgexfoundry/app-functions-sdk-go/blob/master/pkg/transforms/filter.go) (both `FilterByDeviceName` and `FilterByValueDescriptor`), the filters operate on the `Event` model object and return the same objects (`Event` or nil).  Ideally, since both app services and device services generally share the same interface model (from `go-mod-core-contracts`), it would be the desire to share the same filter functions functions between SDKs and associated services.
```

Comment:
I think a common filtering approach would be great, however I think there's more design work required in order for this to be accomplished, especially with respect to configuration of the filters.

Reply:
Agree - made an additional note that this would be taken up in later releases as an additional consideration and as shared filters and use cases are identified.